File: pa_agent/api/analysis_report.py
"""Presentation-focused analysis report payloads for the Web API."""
from __future__ import annotations

from typing import Any

from pa_agent.records.schema import AnalysisRecord


def build_analysis_report(record: AnalysisRecord) -> dict[str, Any]:
    """Return a stable, frontend-friendly report for an analysis record."""
    stage1 = _as_dict(record.stage1_diagnosis)
    stage2 = _as_dict(record.stage2_decision)
    decision = _decision_dict(stage2)
    terminal = _as_dict(stage2.get("terminal"))
    risk_profile = record.meta.decision_stance
    signal_threshold = record.meta.decision_confidence_threshold

    action = _first_text(
        decision.get("action"),
        decision.get("order_type"),
        stage2.get("action"),
        stage2.get("order_type"),
    ) or "不下单"
    direction = _first_optional_text(
        decision.get("direction"),
        decision.get("order_direction"),
        decision.get("trade_direction"),
        stage2.get("direction"),
    )
    summary = _first_text(
        decision.get("reasoning"),
        stage2.get("reasoning"),
        stage2.get("summary"),
        terminal.get("label"),
    )
    risk = _first_text(
        decision.get("risk_assessment"),
        stage2.get("risk_assessment"),
        stage1.get("risk_warning"),
    )

    return {
        "headline": {
            "action": action,
            "direction": direction,
            "summary": summary,
            "risk": risk,
        },
        "metrics": _metrics(stage1, stage2, decision),
        "decision": {
            "order_type": action,
            "direction": direction,
            "risk_profile": risk_profile,
            "signal_threshold": signal_threshold,
            "entry_price": _first_value(decision.get("entry_price"), decision.get("entry")),
            "stop_loss_price": _first_value(
                decision.get("stop_loss_price"),
                decision.get("stop_loss"),
            ),
            "take_profit_price": _first_value(
                decision.get("take_profit_price"),
                decision.get("take_profit"),
            ),
            "terminal": terminal,
        },
        "evidence_tables": [
            {
                "title": "阶段一闸门判断",
                "rows": _trace_rows(stage1.get("gate_trace")),
            },
            {
                "title": "阶段二决策路径",
                "rows": _trace_rows(stage2.get("decision_trace")),
            },
        ],
        "flows": [
            {
                "title": "逐K线结构流程",
                "items": _flow_items(stage1.get("bar_by_bar_summary")),
            }
        ],
        "probability_blocks": [
            _probability_block("下一根K线预测", stage2.get("next_bar_prediction")),
            _probability_block("下一周期预测", stage2.get("next_cycle_prediction")),
        ],
        "lists": [
            {
                "title": "关键因素",
                "items": _string_items(
                    decision.get("key_factors")
                    or stage2.get("key_factors")
                    or stage1.get("key_signals")
                ),
            },
            {
                "title": "观察点",
                "items": _string_items(
                    decision.get("watch_points")
                    or stage2.get("watch_points")
                    or stage1.get("watch_points")
                ),
            },
        ],
    }


def _as_dict(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}


def _decision_dict(stage2: dict[str, Any]) -> dict[str, Any]:
    nested = stage2.get("decision")
    if isinstance(nested, dict):
        return nested
    decision_keys = {
        "action",
        "order_type",
        "direction",
        "order_direction",
        "trade_direction",
        "entry",
        "entry_price",
        "stop_loss",
        "stop_loss_price",
        "take_profit",
        "take_profit_price",
    }
    if any(key in stage2 for key in decision_keys):
        return stage2
    return {}


def _first_value(*values: Any) -> Any:
    for value in values:
        if value is not None and value != "":
            return value
    return None


def _first_text(*values: Any) -> str:
    value = _first_value(*values)
    return "" if value is None else str(value)


def _first_optional_text(*values: Any) -> str | None:
    text = _first_text(*values)
    return text or None


def _coerce_number(value: Any) -> int | float | None:
    if isinstance(value, bool) or value is None or value == "":
        return None
    if isinstance(value, int):
        return value
    if isinstance(value, float):
        return int(value) if value.is_integer() else value
    if isinstance(value, str):
        try:
            number = float(value)
        except ValueError:
            return None
        return int(number) if number.is_integer() else number
    return None


def _tone_for_number(value: int | float) -> str:
    if value >= 60:
        return "good"
    if value > 0:
        return "warn"
    return "bad"


def _metric(label: str, value: Any, unit: str = "") -> dict[str, Any] | None:
    number = _coerce_number(value)
    if number is None:
        return None
    return {
        "label": label,
        "value": number,
        "unit": unit,
        "tone": _tone_for_number(number) if unit == "%" else "neutral",
    }


def _metrics(
    stage1: dict[str, Any],
    stage2: dict[str, Any],
    decision: dict[str, Any],
) -> list[dict[str, Any]]:
    candidates = [
        _metric(
            "诊断信心",
            _first_value(decision.get("diagnosis_confidence"), stage1.get("diagnosis_confidence")),
            "%",
        ),
        _metric(
            "交易信心",
            _first_value(
                decision.get("trade_confidence"),
                decision.get("confidence"),
                stage2.get("trade_confidence"),
                stage2.get("confidence"),
            ),
            "%",
        ),
        _metric(
            "预估胜率",
            _first_value(decision.get("estimated_win_rate"), stage2.get("estimated_win_rate")),
            "%",
        ),
        _metric(
            "历史胜率",
            _first_value(
                decision.get("historical_win_rate_for_this_setup"),
                stage2.get("historical_win_rate_for_this_setup"),
            ),
            "%",
        ),
        _metric(
            "历史样本",
            _first_value(decision.get("historical_sample_count"), stage2.get("historical_sample_count")),
        ),
        _metric(
            "历史期望R",
            _first_value(decision.get("historical_expectancy_r"), stage2.get("historical_expectancy_r")),
            "R",
        ),
    ]
    return [metric for metric in candidates if metric is not None]


def _trace_rows(value: Any) -> list[dict[str, Any]]:
    if not isinstance(value, list):
        return []
    rows: list[dict[str, Any]] = []
    for item in value:
        if not isinstance(item, dict):
            continue
        rows.append(
            {
                "node_id": _first_text(item.get("node_id")),
                "section": _first_text(item.get("section")),
                "question": _first_text(item.get("question")),
                "answer": _first_text(item.get("answer")),
                "reason": _first_text(item.get("reason")),
                "bar_range": _first_text(item.get("bar_range")),
                "skipped": bool(item.get("skipped", False)),
            }
        )
    return rows


def _flow_items(value: Any) -> list[dict[str, Any]]:
    if not isinstance(value, list):
        return []
    items: list[dict[str, Any]] = []
    for index, item in enumerate(value, start=1):
        if not isinstance(item, dict):
            continue
        value_text = _first_text(
            item.get("value"),
            item.get("bar_type"),
            item.get("signal"),
            item.get("context_effect"),
        )
        detail = _first_text(
            item.get("reason"),
            item.get("detail"),
            item.get("context_effect"),
            item.get("follow_through"),
        )
        items.append(
            {
                "id": _flow_id(item, index),
                "label": _first_text(
                    item.get("role"),
                    item.get("label"),
                    item.get("signal"),
                    item.get("bar_type"),
                ),
                "value": value_text,
                "detail": detail,
                "tone": _flow_tone(value_text, detail),
            }
        )
    return items


def _flow_id(item: dict[str, Any], index: int) -> str:
    explicit = _first_text(item.get("id"), item.get("bar"))
    if explicit:
        return explicit
    seq = item.get("seq")
    if seq is not None and seq != "":
        return f"K{seq}"
    return f"K{index}"


def _flow_tone(value: str, detail: str) -> str:
    text = f"{value} {detail}".lower()
    if any(token in text for token in ("bull", "strong", "strengthen", "follow-through")):
        return "good"
    if any(token in text for token in ("bear", "weak", "fail", "risk", "invalid")):
        return "bad"
    if any(token in text for token in ("wait", "pending", "neutral", "range")):
        return "warn"
    return "neutral"


def _probability_block(title: str, value: Any) -> dict[str, Any]:
    prediction = _as_dict(value)
    probabilities = prediction.get("probabilities")
    items = []
    if isinstance(probabilities, dict):
        items = [
            {"label": str(label), "value": probability}
            for label, probability in probabilities.items()
        ]
    return {
        "title": title,
        "items": items,
        "reasoning": _first_text(prediction.get("reasoning")),
    }


def _string_items(value: Any) -> list[str]:
    if not isinstance(value, list):
        return []
    return [str(item) for item in value if item is not None and item != ""]

File: apps/web/features/analysis/analysis-report.ts
import type { AnalysisDecision, AnalysisRecordPayload, AnalysisReport, AnalysisReportTone } from '../../types/api';

export type ReportMetricModel = {
  label: string;
  value: string;
  tone: AnalysisReportTone;
};

export type ReportFieldModel = {
  label: string;
  value: string;
  tone: AnalysisReportTone;
};

export type ReportEvidenceRowModel = {
  node: string;
  section: string;
  question: string;
  answer: string;
  basis: string;
  reason: string;
  skipped: boolean;
};

export type ReportEvidenceTableModel = {
  title: string;
  rows: ReportEvidenceRowModel[];
};

export type ReportFlowItemModel = {
  id: string;
  label: string;
  value: string;
  detail: string;
  tone: AnalysisReportTone;
};

export type ReportFlowModel = {
  title: string;
  items: ReportFlowItemModel[];
};

export type ReportProbabilityItemModel = {
  label: string;
  value: number;
  tone: AnalysisReportTone;
};

export type ReportProbabilityBlockModel = {
  title: string;
  items: ReportProbabilityItemModel[];
  reasoning: string;
};

export type ReportListModel = {
  title: string;
  items: string[];
};

export type ReportDecisionTreeNodeModel = {
  id: string;
  phase: string;
  question: string;
  answer: string;
  basis: string;
  reason: string;
  skipped: boolean;
  tone: AnalysisReportTone;
};

export type ReportDecisionTreeTerminalModel = {
  node: string;
  outcome: string;
  label: string;
  tone: AnalysisReportTone;
};

export type ReportDecisionTreeModel = {
  nodes: ReportDecisionTreeNodeModel[];
  terminal: ReportDecisionTreeTerminalModel | null;
};

export type AnalysisReportModel = {
  headline: {
    action: string;
    direction: string;
    summary: string;
    risk: string;
  };
  metrics: ReportMetricModel[];
  decisionFields: ReportFieldModel[];
  evidenceTables: ReportEvidenceTableModel[];
  flows: ReportFlowModel[];
  probabilityBlocks: ReportProbabilityBlockModel[];
  lists: ReportListModel[];
  decisionTree: ReportDecisionTreeModel;
  hasContent: boolean;
};

const EMPTY_HEADLINE = {
  action: 'pending',
  direction: 'n/a',
  summary: 'No analysis report returned yet.',
  risk: '',
};

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value);
}

function isNonEmpty(value: unknown): boolean {
  return value !== null && value !== undefined && value !== '';
}

function toRecord(value: unknown): Record<string, unknown> {
  return isRecord(value) ? value : {};
}

function toText(value: unknown, fallback = 'n/a'): string {
  if (!isNonEmpty(value)) {
    return fallback;
  }
  return formatReportValue(value);
}

function readDecisionBody(decision?: AnalysisDecision | null): Record<string, unknown> {
  if (!decision) {
    return {};
  }
  const nested = decision.decision;
  if (isRecord(nested)) {
    return { ...decision, ...nested };
  }
  return decision;
}

function readNumber(value: unknown): number | null {
  if (typeof value === 'number' && Number.isFinite(value)) {
    return value;
  }
  if (typeof value === 'string' && value.trim()) {
    const parsed = Number(value.replace('%', '').trim());
    return Number.isFinite(parsed) ? parsed : null;
  }
  return null;
}

function clampPercent(value: number): number {
  return Math.max(0, Math.min(100, Math.round(value)));
}

function toneFromValue(value: unknown): AnalysisReportTone {
  const text = String(value ?? '').toLowerCase();
  if (text.includes('bad') || text.includes('bear') || text.includes('short') || text.includes('risk') || text.includes('weak')) {
    return 'bad';
  }
  if (text.includes('warn') || text.includes('wait') || text.includes('neutral') || text.includes('pending')) {
    return 'warn';
  }
  if (text.includes('good') || text.includes('bull') || text.includes('long') || text.includes('trade') || text.includes('strong')) {
    return 'good';
  }
  return 'info';
}

function metricTone(value: unknown, fallback: AnalysisReportTone = 'neutral'): AnalysisReportTone {
  const numeric = readNumber(value);
  if (numeric === null) {
    return fallback;
  }
  if (numeric >= 70) {
    return 'good';
  }
  if (numeric >= 45) {
    return 'warn';
  }
  return 'bad';
}

function formatMetricValue(value: unknown, unit?: string): string {
  const base = formatReportValue(value);
  if (!unit || base === 'n/a') {
    return base;
  }
  return `${base}${unit}`;
}

function percentMetric(value: unknown): string {
  if (!isNonEmpty(value)) {
    return 'n/a';
  }
  const text = String(value);
  return text.includes('%') ? text : `${formatReportValue(value)}%`;
}

function normalizeEvidenceRows(rows: unknown): ReportEvidenceRowModel[] {
  if (!Array.isArray(rows)) {
    return [];
  }
  return rows.filter(isRecord).map((row) => ({
    node: toText(row.node_id),
    section: toText(row.section, ''),
    question: toText(row.question, ''),
    answer: toText(row.answer, ''),
    basis: toText(row.bar_range, ''),
    reason: toText(row.reason, ''),
    skipped: row.skipped === true,
  }));
}

function normalizeEvidenceTable(title: string, rows: unknown): ReportEvidenceTableModel | null {
  const normalizedRows = normalizeEvidenceRows(rows);
  if (normalizedRows.length === 0) {
    return null;
  }
  return { title, rows: normalizedRows };
}

function normalizeFlowTone(value: unknown): AnalysisReportTone {
  const tone = String(value ?? '').toLowerCase();
  if (tone === 'good' || tone === 'warn' || tone === 'bad' || tone === 'info' || tone === 'neutral') {
    return tone;
  }
  return toneFromValue(value);
}

function traceTone(row: ReportEvidenceRowModel): AnalysisReportTone {
  if (row.skipped) {
    return 'neutral';
  }
  const text = `${row.answer} ${row.reason}`.toLowerCase();
  if (text.includes('yes') || text.includes('是') || text.includes('trade') || text.includes('long') || text.includes('bull')) {
    return 'good';
  }
  if (text.includes('no') || text.includes('否') || text.includes('reject') || text.includes('bear') || text.includes('risk')) {
    return 'bad';
  }
  if (text.includes('wait') || text.includes('hold') || text.includes('neutral') || text.includes('等待') || text.includes('中性')) {
    return 'warn';
  }
  return 'info';
}

function terminalTone(terminal: Record<string, unknown>): AnalysisReportTone {
  const outcome = String(terminal.outcome ?? terminal.label ?? '').toLowerCase();
  if (outcome.includes('trade') || outcome.includes('proceed')) {
    return 'good';
  }
  if (outcome.includes('reject')) {
    return 'bad';
  }
  if (outcome.includes('wait') || outcome.includes('hold')) {
    return 'warn';
  }
  return 'info';
}

function normalizeTerminal(value: unknown): ReportDecisionTreeTerminalModel | null {
  if (!isRecord(value) || Object.keys(value).length === 0) {
    return null;
  }
  return {
    node: toText(value.node_id ?? value.node, ''),
    outcome: toText(value.outcome, ''),
    label: toText(value.label, ''),
    tone: terminalTone(value),
  };
}

function buildDecisionTree(
  tables: ReportEvidenceTableModel[],
  terminal: unknown,
): ReportDecisionTreeModel {
  const nodes = tables.flatMap((table) =>
    table.rows.map((row) => ({
      id: row.node,
      phase: table.title,
      question: row.question,
      answer: row.answer,
      basis: row.basis,
      reason: row.reason,
      skipped: row.skipped,
      tone: traceTone(row),
    })),
  );

  return {
    nodes,
    terminal: normalizeTerminal(terminal),
  };
}

function normalizeProbabilityItems(source: unknown): ReportProbabilityItemModel[] {
  if (Array.isArray(source)) {
    return source
      .filter(isRecord)
      .map((item) => {
        const value = readNumber(item.value);
        return value === null
          ? null
          : {
              label: toText(item.label),
              value: clampPercent(value),
              tone: toneFromValue(item.label),
            };
      })
      .filter((item): item is ReportProbabilityItemModel => Boolean(item));
  }

  if (!isRecord(source)) {
    return [];
  }

  return Object.entries(source)
    .map(([label, value]) => {
      const numeric = readNumber(value);
      return numeric === null
        ? null
        : {
            label,
            value: clampPercent(numeric),
            tone: toneFromValue(label),
          };
    })
    .filter((item): item is ReportProbabilityItemModel => Boolean(item));
}

function listFromValues(title: string, values: unknown[]): ReportListModel | null {
  const items = values.flatMap((value) => {
    if (Array.isArray(value)) {
      return value.map(formatReportValue);
    }
    return isNonEmpty(value) ? [formatReportValue(value)] : [];
  }).filter((item) => item !== 'n/a');
  return items.length > 0 ? { title, items } : null;
}

function compact<T>(items: Array<T | null>): T[] {
  return items.filter((item): item is T => Boolean(item));
}

export function formatReportValue(value: unknown): string {
  if (value === null || value === undefined || value === '') {
    return 'n/a';
  }
  if (typeof value === 'number') {
    return Number.isInteger(value) ? String(value) : String(Number(value.toFixed(3)));
  }
  if (typeof value === 'boolean') {
    return value ? 'true' : 'false';
  }
  if (typeof value === 'string') {
    return value;
  }
  if (Array.isArray(value)) {
    const parts = value.map(formatReportValue).filter((item) => item !== 'n/a');
    return parts.length > 0 ? parts.join(' / ') : 'n/a';
  }
  if (isRecord(value)) {
    const parts = Object.entries(value)
      .filter(([, item]) => isNonEmpty(item))
      .map(([key, item]) => `${key}: ${formatReportValue(item)}`);
    return parts.length > 0 ? parts.join(' / ') : 'n/a';
  }
  return String(value);
}

export function buildAnalysisReportModel(report?: AnalysisReport | null): AnalysisReportModel {
  if (!report) {
    return {
      headline: EMPTY_HEADLINE,
      metrics: [],
      decisionFields: [],
      evidenceTables: [],
      flows: [],
      probabilityBlocks: [],
      lists: [],
      decisionTree: { nodes: [], terminal: null },
      hasContent: false,
    };
  }

  const decision = report.decision ?? {};
  const evidenceTables = report.evidence_tables
    .map((table) => normalizeEvidenceTable(table.title, table.rows))
    .filter((table): table is ReportEvidenceTableModel => Boolean(table));
  const flows = report.flows
    .map((flow) => ({
      title: flow.title,
      items: flow.items.map((item) => ({
        id: toText(item.id),
        label: toText(item.label, ''),
        value: toText(item.value, ''),
        detail: toText(item.detail, ''),
        tone: normalizeFlowTone(item.tone ?? item.value ?? item.label),
      })),
    }))
    .filter((flow) => flow.items.length > 0);
  const probabilityBlocks = report.probability_blocks
    .map((block) => ({
      title: block.title,
      items: normalizeProbabilityItems(block.items),
      reasoning: toText(block.reasoning, ''),
    }))
    .filter((block) => block.items.length > 0);
  const lists = report.lists
    .map((list) => ({ title: list.title, items: list.items.map(formatReportValue).filter((item) => item !== 'n/a') }))
    .filter((list) => list.items.length > 0);
  const decisionTree = buildDecisionTree(evidenceTables, decision.terminal);

  return {
    headline: {
      action: toText(report.headline?.action),
      direction: 'n/a',
      summary: toText(report.headline?.summary, ''),
      risk: toText(report.headline?.risk, ''),
    },
    metrics: report.metrics.map((metric) => ({
      label: metric.label,
      value: formatMetricValue(metric.value, metric.unit),
      tone: metric.tone ?? metricTone(metric.value),
    })),
    decisionFields: [
      { label: 'Risk profile', value: toText(decision.risk_profile), tone: 'info' },
      { label: 'Signal threshold', value: toText(decision.signal_threshold), tone: 'info' },
      { label: 'Entry', value: toText(decision.entry_price), tone: 'info' },
      { label: 'Stop loss', value: toText(decision.stop_loss_price), tone: 'bad' },
      { label: 'Take profit', value: toText(decision.take_profit_price), tone: 'good' },
      { label: 'Terminal', value: toText(decision.terminal), tone: toneFromValue(decision.terminal) },
    ],
    evidenceTables,
    flows,
    probabilityBlocks,
    lists,
    decisionTree,
    hasContent: true,
  };
}

function fallbackMetrics(stage1: Record<string, unknown>, decision: Record<string, unknown>): ReportMetricModel[] {
  return compact([
    isNonEmpty(decision.trade_confidence ?? decision.confidence)
      ? {
          label: 'Trade confidence',
          value: percentMetric(decision.trade_confidence ?? decision.confidence),
          tone: metricTone(decision.trade_confidence ?? decision.confidence),
        }
      : null,
    isNonEmpty(decision.diagnosis_confidence ?? stage1.diagnosis_confidence)
      ? {
          label: 'Diagnosis confidence',
          value: percentMetric(decision.diagnosis_confidence ?? stage1.diagnosis_confidence),
          tone: metricTone(decision.diagnosis_confidence ?? stage1.diagnosis_confidence),
        }
      : null,
    isNonEmpty(decision.estimated_win_rate ?? decision.historical_win_rate_for_this_setup)
      ? {
          label: 'Win rate',
          value: percentMetric(decision.estimated_win_rate ?? decision.historical_win_rate_for_this_setup),
          tone: metricTone(decision.estimated_win_rate ?? decision.historical_win_rate_for_this_setup),
        }
      : null,
    isNonEmpty(decision.historical_sample_count)
      ? { label: 'Samples', value: formatReportValue(decision.historical_sample_count), tone: 'info' }
      : null,
    isNonEmpty(decision.historical_expectancy_r)
      ? { label: 'Expectancy R', value: formatReportValue(decision.historical_expectancy_r), tone: metricTone(decision.historical_expectancy_r, 'info') }
      : null,
  ]);
}

function fallbackDecisionFields(
  decision: Record<string, unknown>,
  stage2: Record<string, unknown>,
  record?: AnalysisRecordPayload | null,
): ReportFieldModel[] {
  return [
    { label: 'Risk profile', value: toText(decision.risk_profile ?? record?.decision_stance), tone: 'info' },
    { label: 'Signal threshold', value: toText(decision.signal_threshold ?? record?.decision_confidence_threshold), tone: 'info' },
    { label: 'Entry', value: toText(decision.entry_price ?? decision.entry), tone: 'info' },
    { label: 'Stop loss', value: toText(decision.stop_loss_price ?? decision.stop_loss), tone: 'bad' },
    { label: 'Take profit', value: toText(decision.take_profit_price ?? decision.take_profit), tone: 'good' },
    { label: 'Terminal', value: toText(stage2.terminal), tone: toneFromValue(stage2.terminal) },
  ];
}

function fallbackEvidenceTables(stage1: Record<string, unknown>, stage2: Record<string, unknown>): ReportEvidenceTableModel[] {
  return compact([
    normalizeEvidenceTable('Decision trace', stage2.decision_trace),
    normalizeEvidenceTable('Diagnostic gate', stage1.gate_trace),
  ]);
}

function fallbackFlows(stage1: Record<string, unknown>): ReportFlowModel[] {
  const rows = stage1.bar_by_bar_summary;
  if (!Array.isArray(rows)) {
    return [];
  }

  const items = rows.filter(isRecord).map((row, index) => {
    const context = toText(row.context_effect, '');
    const reason = toText(row.reason, '');
    const detail = [context, reason].filter(Boolean).join(' / ');
    return {
      id: toText(row.bar, `K${index + 1}`),
      label: toText(row.role, ''),
      value: toText(row.bar_type, ''),
      detail,
      tone: toneFromValue(row.context_effect ?? row.bar_type ?? row.role),
    };
  });

  return items.length > 0 ? [{ title: 'Bar-by-bar flow', items }] : [];
}

function probabilityBlock(title: string, prediction: unknown): ReportProbabilityBlockModel | null {
  if (!isRecord(prediction)) {
    return null;
  }
  const items = normalizeProbabilityItems(prediction.probabilities);
  if (items.length === 0) {
    return null;
  }
  return {
    title,
    items,
    reasoning: toText(prediction.reasoning, ''),
  };
}

function fallbackProbabilityBlocks(stage2: Record<string, unknown>): ReportProbabilityBlockModel[] {
  return compact([
    probabilityBlock('Next bar prediction', stage2.next_bar_prediction),
    probabilityBlock('Next cycle prediction', stage2.next_cycle_prediction),
  ]);
}

function fallbackLists(stage1: Record<string, unknown>, decision: Record<string, unknown>): ReportListModel[] {
  return compact([
    listFromValues('Key factors', [decision.key_factors]),
    listFromValues('Watch points', [decision.watch_points]),
    listFromValues('Observed signals', [stage1.key_signals, stage1.detected_patterns]),
  ]);
}

export function reportFromAnalysisRecord(record?: AnalysisRecordPayload | null, topLevelReport?: AnalysisReport | null): AnalysisReportModel {
  const explicitReport = topLevelReport ?? record?.analysis_report;
  if (explicitReport) {
    return buildAnalysisReportModel(explicitReport);
  }

  const stage1 = toRecord(record?.stage1_diagnosis);
  const stage2 = toRecord(record?.stage2_decision);
  const decision = readDecisionBody(record?.stage2_decision);
  const summaryParts = [
    isNonEmpty(stage1.cycle_position) ? `Cycle: ${formatReportValue(stage1.cycle_position)}` : '',
    isNonEmpty(decision.reasoning) ? formatReportValue(decision.reasoning) : '',
  ].filter(Boolean);
  const metrics = fallbackMetrics(stage1, decision);
  const evidenceTables = fallbackEvidenceTables(stage1, stage2);
  const flows = fallbackFlows(stage1);
  const probabilityBlocks = fallbackProbabilityBlocks(stage2);
  const lists = fallbackLists(stage1, decision);
  const decisionTree = buildDecisionTree(evidenceTables, stage2.terminal ?? decision.terminal);
  const hasContent =
    Object.keys(stage1).length > 0 ||
    Object.keys(stage2).length > 0 ||
    Object.keys(decision).length > 0 ||
    metrics.length > 0 ||
    evidenceTables.length > 0 ||
    flows.length > 0 ||
    probabilityBlocks.length > 0 ||
    lists.length > 0;

  if (!hasContent) {
    return buildAnalysisReportModel(null);
  }

  return {
    headline: {
      action: toText(decision.action ?? decision.order_type, 'decision'),
      direction: 'n/a',
      summary: summaryParts.join(' / '),
      risk: toText(decision.risk_assessment ?? stage1.risk_warning, ''),
    },
    metrics,
    decisionFields: fallbackDecisionFields(decision, stage2, record),
    evidenceTables,
    flows,
    probabilityBlocks,
    lists,
    decisionTree,
    hasContent,
  };
}
